### PR TITLE
Feature/e UI 4353 select location a

### PR DESCRIPTION
**Pull request by paulhowes-HMCTS**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/EUI-4353

### Change description ###
Select a location functionality (including merge from feature/EUI-4690-location-api-to-master)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by Mo-Lala84 — APPROVED** [review]
approved

File: api/locations/index.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * @description getLocations from service ID/location type/search term
+ * @overview API sample: /api/locations/getLocations?serviceIds=SSCS,IA&locationType=hearing&searchTerm=CT91RL
+ * @example service = SSCS | SSCS,IA split with ','
+ * @example locationType = optional | hearing | case_management
+ * @example searchTerm = any search term for postcode | site name | venue name |court name | court address etc.
+ */
+export async function getLocations(req: EnhancedRequest, res: Response, next: NextFunction) {
+  // @ts-ignore
+  const searchTerm = req.query.searchTerm;
```
Comment (paddybasi) [inline api/locations/index.ts:24]: you could "spread" this so it's a concise one line:
`const { searchTerm, serviceIds, locationType } = req.query;`

**Review by tarikdogan-hmcts — APPROVED** [review]

File: api/locations/index.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+function getCourtTypeIdsByService(serviceIdArray: string[]): string {
+  const courtTypeIdsArray = serviceIdArray.map(serviceId => SERVICES_COURT_TYPE_MAPPINGS[serviceId])
+    .reduce(concatCourtTypeWithoutDuplicates);
+  if (courtTypeIdsArray) {
+    return courtTypeIdsArray.join(',');
+  }
+  return '';
+}
+
+function concatCourtTypeWithoutDuplicates(array1: number[], array2: number[]) {
```
Comment (paddybasi) [inline api/locations/index.ts:54]: This looks to be a generic function, possibly could move and rename

File: api/locations/location.mock.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/* tslint:disable:max-line-length */
+import MockAdapter from 'axios-mock-adapter';
+import {HttpMockAdapter} from '../common/httpMockAdapter';
+import {ALL_COURT_LOCATIONS} from './data/location.mock.data';
+import {toEpimsLocation} from "./models/location.model";
```
Comment (paddybasi) [inline api/locations/location.mock.ts:5]: just needs a bit of consistency, double-quotes really should've been picked up by your IDE 😅

File: api/locations/location.mock.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      searchTerm = requestURL.substring(searchTermStartIdx + 1, searchTermEndIdx);
+    } else {
+      searchTerm = requestURL.substring(searchTermStartIdx + 1);
+    }
+    const courTypeIdStartIdx = config.url.lastIndexOf('=');
+    const courTypeIdsStr = requestURL.substring(courTypeIdStartIdx + 1);
+    const searchResult = ALL_COURT_LOCATIONS
+      .filter(locationWithCourtType => courTypeIdsStr.indexOf(locationWithCourtType.court_type_id) !== -1)
+      .filter(location =>
+        (location.court_name ? location.court_name.toLowerCase().includes(searchTerm.toLowerCase()) : false)
```
Comment (paddybasi) [inline api/locations/location.mock.ts:27]: I know it's only a mock, but this is a huge OR clause. Could you not have just put these values into a map/array beforehand and then check to see if searchTerm is in that?

File: src/booking/containers/booking-check/booking-check.component.html [hunk 5]
```diff
@@ -7,7 +7,7 @@ <h1 class="govuk-heading-l">
   <div class="govuk-summary-list__row">
     <dt class="govuk-summary-list__key">Location</dt>
     <dd class="govuk-summary-list__value">
-      <p class="govuk-body">{{bookingProcess.locationName}}</p>
+      <p class="govuk-body">{{bookingProcess.location.court_name}}</p>
```
Comment (paddybasi) [inline src/booking/containers/booking-check/booking-check.component.html:10]: Not sure if this is by design, but I'd prefer the camelCase consistency

**Review by paddybasi — CHANGES_REQUESTED** [review]

File: api/locations/data/serviceCourtType.mapping.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+export const SERVICES_COURT_TYPE_MAPPINGS = {
+  EMPLOYMENT: [17], // 17: Employment Tribunal
```
[the remark this answers is not in the document]
Comment (paulhowes-HMCTS) [reply api/locations/data/serviceCourtType.mapping.ts:2]: Merged from https://github.com/hmcts/rpx-xui-webapp/tree/feature/EUI-4690-location-api-to-master - not part of the changes in this PR.

File: api/locations/index.ts (hunk 1 above)
Comment (paulhowes-HMCTS) [reply api/locations/index.ts:24]: Merged from https://github.com/hmcts/rpx-xui-webapp/tree/feature/EUI-4690-location-api-to-master - not part of the changes in this PR.

File: api/locations/location.mock.ts (hunk 3 above)
Comment (paulhowes-HMCTS) [reply api/locations/location.mock.ts:5]: Merged from https://github.com/hmcts/rpx-xui-webapp/tree/feature/EUI-4690-location-api-to-master - not part of the changes in this PR.

File: api/locations/location.mock.ts (hunk 4 above)
Comment (paulhowes-HMCTS) [reply api/locations/location.mock.ts:27]: Merged from https://github.com/hmcts/rpx-xui-webapp/tree/feature/EUI-4690-location-api-to-master - not part of the changes in this PR.

File: src/booking/containers/booking-check/booking-check.component.html (hunk 5 above)
Comment (paulhowes-HMCTS) [reply src/booking/containers/booking-check/booking-check.component.html:10]: I'd definitely prefer it too, however the location model is using LocationByEPIMSModel in the common lib and this is the structure returned from the API endpoint.

**Review by paddybasi — APPROVED** [review]

**Pull request merged**